### DISCUSSION TITLE
run directus with pm2 as a daemon/service

### DISCUSSION THREAD

**Discussion by yii-joblo** [directus/directus#3254] [Archive]
1) install pm2: 

`npm install pm2 -g`

2) edit package.json of your directus-installation and add the start script: **npx directus start**

```json
{
  "name": "directus",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "npx directus start",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@directus/sdk-js": "^9.0.0-rc.19",
    "directus": "^9.0.0-rc.8",
    "mysql": "^2.18.1"
  }
}
```

3) run in your installation path of directus:
`pm2 start npm --name "directus" -- start`

directus is online:

`http://localhost:8055/`

Now you can control the service
see  [https://pm2.keymetrics.io/]( https://pm2.keymetrics.io/)

pm2 stop|restart|delete directus
pm2 monit
pm2 logs
...

**Comment by 7ochukwu** [comment]
tnx a lot!

**Reply by RageBrahimi** [reply]
FYI, more [here](https://docs.directus.io/guides/installation/ubuntu/#ubuntu)

**Reply by supermx1** [reply]
This isn't working. It shows the status as `stopped`

**Reply by rijkvanzanten** [reply]
Where does it show that status, what is logged, how did you set it up, what did you try so far, where are you running this, etc etc

**Reply by supermx1** [reply]
It just went crazy after updating to rc95. I initially created a startup file, which worked great for a while. Then I needed to scale my instance to multiple clusters. So I ran a pm2 script that worked and split my instance into 3. Now, after the update, everything has gone to hell. It starts and immediately stops with a stopped or errored status in pm2. If I use the script which was my old method, it has a status of running but if I check the logs it is full of errors and directus doesn't start.
 
 I am running it on my Windows Server Machine (Windows Server 2019).

**Reply by supermx1** [reply]
This is my PM2 log

 [TAILING] Tailing last 15 lines for [all] processes (change the value with --lines option)
C:\Users\Administrator\.pm2\pm2.log last 15 lines:
PM2        | 2021-09-29T02:37:01: PM2 log: App [npm:0] online
PM2        | 2021-09-29T02:37:01: PM2 log: App [npm:0] exited with code [1] via signal [SIGINT]
PM2        | 2021-09-29T02:37:01: PM2 log: App [npm:0] starting in -fork mode-
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] online
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] exited with code [1] via signal [SIGINT]
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] starting in -fork mode-
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] online
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] exited with code [1] via signal [SIGINT]
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] starting in -fork mode-
PM2        | 2021-09-29T02:37:02: PM2 log: App [npm:0] online
PM2        | 2021-09-29T02:37:03: PM2 log: App [npm:0] exited with code [1] via signal [SIGINT]
PM2        | 2021-09-29T02:37:03: PM2 log: App [npm:0] starting in -fork mode-
PM2        | 2021-09-29T02:37:03: PM2 log: App [npm:0] online
PM2        | 2021-09-29T02:37:03: PM2 log: App [npm:0] exited with code [1] via signal [SIGINT]
PM2        | 2021-09-29T02:37:03: PM2 log: Script C:\PROGRAM FILES\NODEJS\NPM.CMD had too many unstable restarts (16). Stopped. "errored"

C:\Users\Administrator\.pm2\logs\npm-out.log last 15 lines:
C:\Users\Administrator\.pm2\logs\npm-error.log last 15 lines:
0|npm      | C:\PROGRAM FILES\NODEJS\NPM.CMD:1
0|npm      | :: Created by npm, please don't edit manually.
0|npm      | ^
0|npm      |
0|npm      | SyntaxError: Unexpected token ':'
0|npm      |     at wrapSafe (internal/modules/cjs/loader.js:1001:16)
0|npm      |     at Module._compile (internal/modules/cjs/loader.js:1049:27)
0|npm      |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
0|npm      |     at Module.load (internal/modules/cjs/loader.js:950:32)
0|npm      |     at Function.Module._load (internal/modules/cjs/loader.js:790:14)
0|npm      |     at Object.<anonymous> (C:\Users\Administrator\AppData\Roaming\npm\node_modules\pm2\lib\ProcessContainerFork.js:33:23)
0|npm      |     at Module._compile (internal/modules/cjs/loader.js:1085:14)
0|npm      |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
0|npm      |     at Module.load (internal/modules/cjs/loader.js:950:32)
0|npm      |     at Function.Module._load (internal/modules/cjs/loader.js:790:14)

**Reply by supermx1** [reply]
Here is a video

https://user-images.githubusercontent.com/9326922/135189704-cf8be5dd-b9cc-4a67-9d73-171a262a2d11.mp4

**Reply by azrikahar** [reply]
Seems to be a known pm2 issue for windows: https://github.com/Unitech/pm2/issues/3657#issuecomment-926890502

Perhaps you can try the solution offered there (above link)? Or this one which should be similar but simpler than it: https://stackoverflow.com/a/56203967

**Reply by supermx1** [reply]
Amazing @azrikahar. Good work as always. The fix for Windows server is to create a js file and paste the following inside.


var exec = require('child_process').exec;
exec('npx directus start', {windowsHide: true});


Save it and then run.

pm2 start <name of js file>.js 

eg `pm2 start rundirectus.js`

**Reply by supermx1** [reply]
I wonder if we can include this in the documentation @rijkvanzanten ? So users of Windows won't suffer like I did. Lol

**Reply by rijkvanzanten** [reply]
Sure! Wanna write a little "How to run with PM2" guide @supermx1?

**Reply by licitdev** [reply]
@supermx1 I have been using PM2 on macOS and linux servers without any issue.
So the issue occurs only on Windows.

**Reply by HemboPagi** [reply]
Hello

I managed to get it working nicely, but pm2 keeps restarting directus almost every minute. CPU shows in PM2 monitor 100% before the start. Is it about the virtual machine configuration issue, lack of resources? Any thoughts?

**Reply by jamescammarano** [reply]
> Hello
> 
> I managed to get it working nicely, but pm2 keeps restarting directus almost every minute. CPU shows in PM2 monitor 100% before the start. Is it about the virtual machine configuration issue, lack of resources? Any thoughts?

What OS are you on and what are your resources?

**Reply by HemboPagi** [reply]
Hello and thank you for responding!

I've started PM2 with following parameters (ecosystem.config.js ):

module.exports = {
  apps : [{
    name   : "directus",
    instances : "2",
    exec_mode : "cluster",
    script : "npx",
    args : "directus start",
    autorestart : false,
    watch: false,
    max_memory_restart: "2G"
  }]
}

Memory 8GiB RAM, 8Gib Swap , 2 CPU

Linux version 4.18.0-348.12.2.el8_5.x86_64 (mockbuild@dal1-prod-builder001.bld.equ.rockylinux.org) (gcc version 8.5.0 20210514 (Red Hat 8.5.0-3) (GCC)) #1 SMP Wed Jan 19 17:53:40 UTC 2022

**Reply by joselcvarela** [reply]
If it's restarting often, I believe it could be cause by some startup issue like migrations or extensions;
Can you see logs? Also you can set env var `DEBUG=*` in order to see what's happening.

**Reply by HemboPagi** [reply]
I can see logs but it does not speak to me :)

I can see that knex:bindings is undefined in several places:
`knex:bindings [ 'directus' ] undefined`
`knex:bindings [ 'active', 100 ] undefined`

<details>
<summary>
Here is a small excerpt :
</summary>

0|directus  | 2022-02-21T12:17:08.135Z follow-redirects options {
0|directus  |   maxRedirects: 21,
0|directus  |   maxBodyLength: 10485760,
0|directus  |   protocol: 'https:',
0|directus  |   path: '/',
0|directus  |   method: 'POST',
0|directus  |   headers: {
0|directus  |     Accept: 'application/json, text/plain, */*',
0|directus  |     'Content-Type': 'application/json',
0|directus  |     'User-Agent': 'axios/0.24.0',
0|directus  |     'Content-Length': 542
0|directus  |   },
0|directus  |   agent: undefined,
0|directus  |   agents: { http: undefined, https: undefined },
0|directus  |   auth: undefined,
0|directus  |   hostname: 'telemetry.directus.io',
0|directus  |   port: null,
0|directus  |   nativeProtocols: {
0|directus  |     'http:': {
0|directus  |       _connectionListener: [Function: connectionListener],
0|directus  |       METHODS: [Array],
0|directus  |       STATUS_CODES: [Object],
0|directus  |       Agent: [Function],
0|directus  |       ClientRequest: [Function: ClientRequest],
0|directus  |       IncomingMessage: [Function: IncomingMessage],
0|directus  |       OutgoingMessage: [Function: OutgoingMessage],
0|directus  |       Server: [Function: Server],
0|directus  |       ServerResponse: [Function: ServerResponse],
0|directus  |       createServer: [Function: createServer],
0|directus  |       validateHeaderName: [Function: __node_internal_],
0|directus  |       validateHeaderValue: [Function: __node_internal_],
0|directus  |       get: [Function: get],
0|directus  |       request: [Function: request],
0|directus  |       maxHeaderSize: [Getter],
0|directus  |       globalAgent: [Getter/Setter]
0|directus  |     },
0|directus  |     'https:': {
0|directus  |       Agent: [Function: Agent],
0|directus  |       globalAgent: [Agent],
0|directus  |       Server: [Function: Server],
0|directus  |       createServer: [Function: createServer],
0|directus  |       get: [Function: get],
0|directus  |       request: [Function: request]
0|directus  |     }
0|directus  |   }
0|directus  | }
	0|directus  | 14:18:33 ✨ Shutting down...
0|directus  | 14:18:33 ✨ Database connections destroyed
0|directus  | 14:18:33 ✨ Directus shut down OK. Bye bye!
1|directus  | 14:18:33 ✨ Shutting down...
PM2         | pm2 has been killed by signal, dumping process list before exit...
PM2         | Deleting process 0
PM2         | Stopping app:directus id:0
PM2         | App name:directus id:0 disconnected
1|directus  | 14:18:33 ✨ Database connections destroyed
1|directus  | 14:18:33 ✨ Directus shut down OK. Bye bye!



</details>

**Reply by HemboPagi** [reply]
it all seems to do with _knex_. or can it be related with underlying MariaDB?

**Comment by biapar** [comment]
On Ubuntu works.

https://docs.directus.io/getting-started/installation/ubuntu/

**Comment by nexthis** [comment]
for directus ^10 I use config and everything looks good 
Of course, when using a cluster, remember to disable the cache or switch to Redis.
```
module.exports = {
    apps : [{
        name : "directus",
        instances : "max",
        exec_mode : "cluster",
        script : "./node_modules/@directus/api/dist/start.js",
        autorestart : false,
        watch: false,
        max_memory_restart: "2G"
    }]
}
```

**Reply by rijkvanzanten** [reply]
> Of course, when using a cluster, remember to disable the cache or switch to Redis.

~~remember to disable the cache or~~ **switch to Redis**

It's not just caching; it's also rate-limiting, websockets, cron jobs, etc. When running multiple instances that all connect to the same database, redis is required 👍🏻
